src/webextensions.c

```diff
@@ -0,0 +1,629 @@
+          if (!json_parser_load_from_file (parser, metadata_filename, error))
+            return NULL;
+
+          metadata_root = json_node_get_object (json_parser_get_root (parser));
+
+          /* Skip if this isn't the requested native messaging server */
+          if (g_strcmp0 (json_object_get_string_member (metadata_root, "name"), server_name) != 0)
+            continue;
+
+          /* Skip if this is not a "stdio" type native messaging server */
+          if (g_strcmp0 (json_object_get_string_member (metadata_root, "type"), "stdio") != 0)
+            continue;
+
+          /* Skip if this server isn't available to the extension. Note that  */
+          if (!array_contains (json_object_get_array_member (metadata_root, "allowed_extensions"), extension_or_origin) &&
+              !array_contains (json_object_get_array_member (metadata_root, "allowed_origins"), extension_or_origin))
+            continue;
+
+          /* Server matches: return its executable path and description */
+          if (server_description != NULL)
+            *server_description = g_strdup (json_object_get_string_member (metadata_root, "description"));
+          return g_strdup (json_object_get_string_member (metadata_root, "path"));
+        }
+    }
+
+  g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_FOUND, "cannot find native messaging server");
+  return NULL;
+}
+
+static void
+handle_start_in_thread (GTask *task,
+                        gpointer source_object,
+                        gpointer task_data,
+                        GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  Session *session;
+  WebExtensionsSession *we_session;
+  const char *arg_name;
+  const char *arg_extension_or_origin;
+  const char *app_id;
+  g_autofree char *server_path = NULL;
+  g_autofree char *server_description = NULL;
+  guint response = XDG_DESKTOP_PORTAL_RESPONSE_OTHER;
+  gboolean should_close_session;
+  Permission permission;
+  gboolean allowed;
+  char *argv[] = {NULL, NULL};
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+  session = g_object_get_data (G_OBJECT (request), "session");
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+  g_object_set_data (G_OBJECT (request), "session", NULL);
+  we_session = (WebExtensionsSession *)session;
+
+  if (!request->exported)
+    goto out;
+
+  arg_name = g_object_get_data (G_OBJECT (request), "name");
+  arg_extension_or_origin = g_object_get_data (G_OBJECT (request), "extension-or-origin");
+
+  server_path = find_server (arg_name, arg_extension_or_origin, &server_description, &error);
+  if (server_path == NULL)
+    {
+      g_warning ("Could not find WebExtensions backend: %s", error->message);
+      goto out;
+    }
+
+  app_id = xdp_app_info_get_id (request->app_info);
+  permission = get_permission_sync (app_id, PERMISSION_TABLE, arg_name);
+  if (permission == PERMISSION_ASK || permission == PERMISSION_UNSET)
```

Comment:
I'm testing this with the firefox snap (patched with https://phabricator.services.mozilla.com/D140803), where `app_id == "snap.firefox"` in an Ubuntu jammy environment, and I'm not getting prompted for permission, instead the permission defaults to NO, and I have to change it by invoking `flatpak permission-set` to get it working. `xdg-desktop-portal-gtk` and `xdg-desktop-portal-gnome` are running.

Reply:
@jhenstridge any idea on this?

Reply:
@robert-ancell: this was #769. @oSoMoN's patch in this PR gets the access dialog to work for this portal's use cases, but it is a more general problem.